### ISSUE TITLE
Social Login with Passport.js

### ISSUE THREAD

**Issue by dawidvdh** [bertho-zero/react-redux-universal-hot-example#6]
Hey, I posted a similar question on the original project but i'm pretty sure its dead and noticed that you had implemented passportjs on this fork and that this seems a lot more lively so here is the question: 

I am trying to implement a social login such as facebook using [passport.js with the facebook strategy](https://github.com/jaredhanson/passport-facebook).

I am however having some difficulty understanding how to get this to work with this framework and react router, are there any guides/examples that anyone can point me to on how to achieve this?

Any Help would be Greatly Appreciated.

**Comment by jahmaiosullivan** [comment]
Hey dawidvdh, I added FB auth strategy using passport in my project here: https://github.com/jahmaiosullivan/hobbyclue. Here is how it works:


First off as you may already know, the project is split into two parts:

1. Web server running on PORT 3000
2. API server running on PORT 3030 (proxied to /api/ on PORT 3000).


I implemented passport in the API server side. If you check my code inside the file [https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/index.js](https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/index.js) you'll see a call:

`configureAuth(app, config);` on Line 42.

This file is here: [https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/configureAuth.js](https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/configureAuth.js)
This file implements the FB strategy to login. 


Here is the contents of that file for easier reading:

`import passport from 'passport';
import {Strategy as FacebookStrategy} from 'passport-facebook';
import {Strategy as LocalStrategy} from 'passport-local';
import {User, UserClaim, comparePassword} from './database/models';
import jwt from 'jsonwebtoken';
import util from 'util';
import config from '../config';

const configure = (app, config) => {

  function addJWT(user){
    const token = jwt.sign({ email: user.email }, config.auth.jwt.secret, {
      expiresIn: 60000
    });
    return Object.assign({}, user.toJSON(), {token});
  }

  app.use(passport.initialize());
  app.use(passport.session());

  app.get('/login/facebook', passport.authenticate('facebook'));

  app.get('/login/facebook/return',
    passport.authenticate('facebook', {failureRedirect: '/login'}),
    function (req, res) {
      res.redirect(`http://${config.host}:${config.port}`);
    }
  );

  // Authentication
  passport.use(new FacebookStrategy({
      clientID: config.auth.facebook.id,
      clientSecret: config.auth.facebook.secret,
      callbackURL: '/api/login/facebook/return',
      passReqToCallback: true,
      profileFields: ['id', 'emails', 'name', 'displayName', 'gender']
    },
    function (req, accessToken, refreshToken, profile, done) {
      const claimType = 'urn:facebook:access_token';

      const findOrCreateFBUser = async() => {
        if (profile) {
          // Look up user by profile id
          let user = await User.findOne({
            where: {profileType: profile.provider, profileId: profile.id}
          });

          // Create a new user in the user table if not found
          if (!user) {
            var newUser = {
              name: profile.displayName,
              email: profile.emails[0].value,
              profileId: profile.id,
              profileType: profile.provider,
              gender: profile.gender,
              picture: `https://graph.facebook.com/${profile.id}/picture?type=large`,
              claims: [
                {type: claimType, value: profile.id}
              ]
            };
            console.log(`New user based on FB profile is ${util.inspect(newUser)}`);
            user = await User.create(newUser, {
              include: [
                {model: UserClaim, as: 'claims'}
              ]
            });
          }

          // Return the user
          const userWithToken = addJWT(user);
          done(null, userWithToken);
        }
      };

      findOrCreateFBUser().catch(done);
    }
  ));

  passport.use(new LocalStrategy({
      usernameField: 'email'
    },
    function (username, password, done) {
      const findUser = async() => {
        // Look up user by email
        let user = await User.findOne({
          where: {email: username}
        });

        if (!user) return done(new Error('Authentication failed. User not found.'));

        if (comparePassword(password, user.passwordHash)) {
          done(null, addJWT(user));
        }
        else {
          done(new Error(`passwords do not match for user ${username}.`));
        }
      };

      findUser().catch(done);
    }
  ));

  passport.serializeUser(function (user, cb) {
    cb(null, user);
  });

  passport.deserializeUser(function (obj, cb) {
    cb(null, obj);
  });
};

export default configure;`


We need a way to trigger FB login (i.e when a user clicks the "Login with FB" button), so I added a login API path that triggers the FB auth strategy. That is here [https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/actions/auth/login.js](https://github.com/jahmaiosullivan/hobbyclue/blob/master/api/actions/auth/login.js)

Here is the contents of that file for easier reading: 

`import {User} from '../../database/models';
import passport from 'passport';

export default function login( req, res, next ) {
  return new Promise((resolve, reject) => {
    passport.authenticate( 'local', function authenticate(err, user, info){
      if(err)    return reject(err);
      if(!user) return reject(info);

      req.logIn(user, err => {
        if(err) { reject( err ); }
        else {
          resolve(user);
        }
      })

    })(req, res, next);
  });
}`


We can't actually trigger FB login client-side since FB needs to redirect back to us, so I simply added an actual real-link to the login page that logs you in via FB server-side. I added it to my actual login page container here [https://github.com/jahmaiosullivan/hobbyclue/blob/master/src/containers/Login/Login.js](https://github.com/jahmaiosullivan/hobbyclue/blob/master/src/containers/Login/Login.js).

The code from that page that is relevant is this:

`<a href="/api/login/facebook">
                <svg
                  width="30"
                  height="30"
                  viewBox="0 0 30 30"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M22 16l1-5h-5V7c0-1.544.784-2 3-2h2V0h-4c-4.072 0-7 2.435-7 7v4H7v5h5v14h6V16h4z"
                  />
                </svg>
                <span>Log in with Facebook</span>
              </a>`



So to recap what we did:

1. Configured passport's FB auth strategy in the API server  (wired into Express).
2. Added an /api/auth/login route that triggers API auth strategy
3. Added a FB link to the Login page
4. Done!!!!


P.S: Remember to set your FB APP ID and secret.
In my case, I am passing it via config to my configureAuth function (Step 1 above).

**Comment by dawidvdh** [comment]
hey @jahmaiosullivan,

thanks so much for all your effort, really do appreciate it... works perfectly, there's only one part where i am left a little bit confused still because i need more information than facebook provides, so i need to send the user data from facebook that the api provides to the redux store, Would you have any idea on how to achieve this? Last question i promise...

**Comment by bertho-zero** [comment]
Hi @jahmaiosullivan,

It is also possible to connect with the SDK JS Facebook and use passport-facebook-token with access_token.

As one does not leave the application to connect directly to the API, the connection is made in a new window, you can call a callback `onLogin` to request the API with accessToken, for summary:

1. button login FB with the SDK (or [react-facebook-login](https://github.com/keppelen/react-facebook-login/), the code is easy to read) with callback
2. callback with facebook response, if logged you send access_token (or more) to your api
3. API: in passport-facebook-token Strategy you create or update user, sign jwt and return user

**Comment by bertho-zero** [comment]
You can watch and try my [social-login](https://github.com/bertho-zero/react-redux-universal-hot-example/tree/social-login) branch.

I created `FacebookLogin` component for use the Facebook JS SDK easier, I added `passport-facebook` & `passport-facebook-token` to Feathers for verify token, create/update the user with facebook data on server, sign jwt and return user.

**Comment by bertho-zero** [comment]
I just added the connection synchronization between rest and websocket, now you can connect on local/token/oauth in rest and use sockets in secure way with feathers :smiley:

**Comment by headwinds** [comment]
@bertho-zero Last night I attempted to hook up passport - I got stuck on the port configuration and keep getting cross domain issues even though I was using cors. I'm attempting to upgrade an app that doesn't use this proxy API setup so I know my passport setup works without the proxy. 

When you go into setup for your Facebook or Google test app for localhost, what is the domain considering the port switch from 3000 to 3030 through this proxy.

Should I setup my test app with:

http://localhost:3000/auth/facebook

or 

http://localhost:3030/auth/facebook

And then same thing for the callback URLs. What do your localhost urls look like in Facebook?

**Comment by bertho-zero** [comment]
The proxy prefix the principal port with '/api', it should be `http://localhost:3000/api/auth/facebook`.

**Comment by githubitsme** [comment]
It is easier to implement facebook authentication compare to twitter authentication.Because twitter use session Oauth 1.0 It need Session.So how to implement twitter strategy with passport redux

**Comment by advance512** [comment]
Hi,

What happened to the Social Auth branch? The link is dead. Is the code now in feathers-authentication-v1?

Thanks!

**Comment by BrennerSpear** [comment]
Have the same question as @githubitsme - can this work the same way for Twitter?
